Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,667 @@
+    ///    method with the decoded event data.
+    /// 3. Return the encoded output of (2) with `self.encoder`
+    public mutating func handle(
+        _ request: ByteBuffer,
+        context: LambdaContext
+    ) async throws -> LambdaResponse
+}
+```
+
+### Handler as a Closure
+
+To create a Lambda function using the current API, a user first has to create an object and conform it to one of the
+handler protocols by implementing the initializer and the `handle(...)` function. Now that `LambdaRuntime` is public,
+this verbosity can very easily be simplified.
+
+#### ClosureHandler
+
+This handler is generic over any `Event` type conforming to `Decodable` and any `Output` type conforming to `Encodable`
+or `Void`.
+
+```swift
+public struct ClosureHandler<Event, Output>: LambdaHandler {
+    /// Initialize with a closure handler over generic Input and Output types
+    public init(body: @escaping (Event, LambdaContext) async throws -> Output) where Output: Encoda
+    /// Initialize with a closure handler over a generic Input type (Void Output).
+    public init(body: @escaping (Event, LambdaContext) async throws -> Void) where Output == Void
+    /// The business logic of the Lambda function.
+    public func handle(_ event: Event, context: LambdaContext) async throws -> Output
+}
+```
+
+Given that `ClosureHandler` conforms to `LambdaHandler`:
+
+1. We can extend the `LambdaRuntime` initializer such that it accepts a closure as an argument.
+2. Within the initializer, the closure handler is wrapped with `LambdaCodableAdapter`.
+
+```swift
+extension LambdaRuntime {
+    /// Initialize a LambdaRuntime with a closure handler over generic Event and Output types.
+    /// This initializer bolts on encoding/decoding logic by wrapping the closure handler with
+    /// LambdaCodableAdapter.
+    public init<Event: Decodable, Output: Encodable>(
+        body: @escaping (Event, LambdaContext) async throws -> Output
+    ) where Handler == LambdaCodableAdapter<ClosureHandler<Event, Output>, Event, Output, JSONDecoder, JSONEncoder>
+
+    /// Same as above but for handlers with a void output
+    public init<Event: Decodable>(
+        body: @escaping (Event, LambdaContext) async throws -> Void
+    ) where Handler == LambdaCodableAdapter<ClosureHandler<Event, Void>, Event, Void, JSONDecoder, VoidEncoder>
+}
+```
+
+We can now significantly reduce the verbosity and leverage Swift's trailing closure syntax to cleanly create and run a
+Lambda function, abstracting away the decoding and encoding logic from the user:
+
+```swift
+/// The type the handler will use as input
+struct Input: Decodable {
+    var message: String
+}
+
+/// The type the handler will output
+struct Greeting: Encodable {
+    var echoedMessage: String
+}
+
+/// A simple Lambda function that echoes the input
+let runtime = LambdaRuntime { (event: Input, context: LambdaContext) in
+    Greeting(echoedMessage: event.message)
+}
+
+try await runtime.run()
```

Comment:
For this, the simplest use case, would it make sense to provide a static `main()` on `LambdaRuntime` that just calls its `run` method so users can use `@main` if this is all they need to do?

Reply:
Wouldn't it look like-

```swift
@main
LambdaRuntime { (event: Input, context: LambdaContext) in
    Greeting(echoedMessage: event.message)
}
```